addons/website_sale/controllers/main.py

```diff
@@ -245,21 +243,30 @@ def shop(self, page=0, category=None, search='', ppg=False, **post):
         if attrib_list:
             post['attrib'] = attrib_list
 
+        domain = self._get_search_domain(search, attrib_values)
+
+        Category.flush()
         Product = request.env['product.template'].with_context(bin_size=True)
+        where, args = expression.expression(domain, Product).to_sql()
+        query = """
+            SELECT c.id, COUNT(product_template)
+            FROM product_template, product_public_category c, product_public_category_product_template_rel
+            WHERE product_template_id = product_template.id AND
+            (SELECT parent_path FROM product_public_category WHERE id = product_public_category_id) LIKE CONCAT(c.parent_path, '%%') AND """ + where + """
+            GROUP BY c.id
+        """
+        request.env.cr.execute(query, args)
+        categs_nbr_prod = dict(request.env.cr.fetchall())
+        categs_ids = list(categs_nbr_prod.keys())
 
-        search_product = Product.search(domain)
         website_domain = request.website.website_domain()
-        categs_domain = [('parent_id', '=', False)] + website_domain
-        if search:
-            search_categories = Category.search([('product_tmpl_ids', 'in', search_product.ids)] + website_domain).parents_and_self
-            categs_domain.append(('id', 'in', search_categories.ids))
-        else:
-            search_categories = Category
-        categs = Category.search(categs_domain)
+        categs = Category.search([('id', 'in', categs_ids)] + website_domain)
```

Comment:
why not apply website domain directly in _get_search_domain?

it will avoid to count product for category from other website (not critical)